File: Kth Ancestor in a Tree[26-July-2023]/Approach1[Using Map]/solution.cpp
//  >> C++ CODE <<


void getParents(Node* root, int par, unordered_map<int, int>& parentMap) {
    if (root == nullptr) {
        return;
    }
    parentMap[root->data] = par;
    getParents(root->left, root->data, parentMap);
    getParents(root->right, root->data, parentMap);
}

int kthAncestor(Node* root, int k, int node) {
    unordered_map<int, int> parentMap;
    getParents(root, -1, parentMap);
    
    int nn = node;
    
    while (parentMap.find(nn)!=parentMap.end() && k > 0) {
        nn = parentMap[nn];
        k--;
    }
    
    return parentMap.find(nn)==parentMap.end() ? -1 : nn;
}

File: Kth Ancestor in a Tree[26-July-2023]/Approach1[Using Map]/Solution.py
#  >> PYTHON CODE <<

def kthAncestor(root,k, node):
    def getParents(root,par,parent_map):
        if(not root): return
    
        parent_map[root]=par
        
        getParents(root.left,root,parent_map)
        getParents(root.right,root,parent_map)
    
    def find(root,node):
        if(not root): return None
        if(root.data==node): return root
        
        left_search=find(root.left,node)
        right_search=find(root.right,node)
        
        return right_search if not left_search else left_search
            
    
    mp={}
    
    getParents(root,-1,mp)
    nn=find(root,node)
    
    while(nn in mp and k>0):
        nn=mp[nn]
        k-=1
    
    return nn.data if nn in mp else -1